crates/codegen/language/definition/src/model/utils/mod.rs

```diff
@@ -1,5 +1,78 @@
+use crate::model::{self, Item};
+use semver::Version;
+use std::collections::BTreeSet;
+
 mod identifier;
 mod version_specifier;
 
 pub use identifier::*;
 pub use version_specifier::*;
+
+/// Collects all versions that changes the language grammar in a breaking way.
+pub fn collect_breaking_versions(lang: &model::Language) -> BTreeSet<Version> {
+    let first = lang.versions.first().cloned().unwrap();
+    let mut res = BTreeSet::from_iter([first]);
+
+    fn add_spec(res: &mut BTreeSet<Version>, spec: &Option<model::VersionSpecifier>) {
+        let Some(spec) = spec else {
+            return;
+        };
+
+        match spec.clone() {
+            model::VersionSpecifier::Never => (),
+            model::VersionSpecifier::From { from } => {
+                res.insert(from);
+            }
+            model::VersionSpecifier::Till { till } => {
+                res.insert(till);
+            }
+            model::VersionSpecifier::Range { from, till } => {
+                res.insert(from);
+                res.insert(till);
+            }
+        }
+    }
+
+    for item in lang
```

Comment:
I suggest creating a language-specific extension, since it seems to be a common pattern:

```rust
impl Language {
  fn items(&self) -> impl Iterator<&Rc<Item>> {}
}
```

Reply:
Done in dc0112de9